File: Leetcode/tree/114_FlattenBinaryTreeToLinkedList/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private TreeNode prev = null;
    public void flatten(TreeNode root) {
        if(root==null) return;
        TreeNode left = root.left, right = root.right;
        flatten(right);
        flatten(left);
        root.right = prev;
        root.left = null;
        prev = root;
    }
}

File: Leetcode/tree/114_FlattenBinaryTreeToLinkedList/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        prev = [None]
        self.dfs(root, prev)

    def dfs(self, root, prev):
        if not root: return None
        left = root.left
        right = root.right
        self.dfs(right, prev)
        self.dfs(left, prev)
        root.right = prev[0]
        root.left = None
        prev[0] = root

